File: src/iOS/LoginViewController.m
//
//  LoginViewController.m
//  Go Map!!
//
//  Created by Bryce Cogswell on 12/19/12.
//  Copyright (c) 2012 Bryce Cogswell. All rights reserved.
//

#import "AppDelegate.h"
#import "EditorMapLayer.h"
#import "KeyChain.h"
#import "LoginViewController.h"
#import "MapView.h"
#import "OsmMapData.h"

@implementation LoginViewController

- (void)viewDidLoad
{
    [super viewDidLoad];
}

- (IBAction)textFieldReturn:(id)sender
{
	[sender resignFirstResponder];
}

- (IBAction)textFieldDidChange:(id)sender
{
	_saveButton.enabled = _username.text.length && _password.text.length;
}

- (IBAction)registerAccount:(id)sender
{
	[[UIApplication sharedApplication] openURL:[NSURL URLWithString: @"https://www.openstreetmap.org/user/new"]];
}


- (IBAction)verifyAccount:(id)sender
{
	if ( _activityIndicator.isAnimating )
		return;
    
    NSString *username = [_username.text stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]];
    NSString *password = [_password.text stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]];

	AppDelegate * appDelegate = (id)[[UIApplication sharedApplication] delegate];
    appDelegate.userName		= username;
    appDelegate.userPassword	= password;

	_activityIndicator.color = UIColor.darkGrayColor;
	[_activityIndicator startAnimating];

	[appDelegate.mapView.editorLayer.mapData verifyUserCredentialsWithCompletion:^(NSString * errorMessage){
		[_activityIndicator stopAnimating];
		if ( errorMessage ) {

			// warn that email addresses don't work
			if ( [appDelegate.userName containsString:@"@"] ) {
				errorMessage = NSLocalizedString(@"You must provide your OSM user name, not an email address.",nil);
			}
			UIAlertController * alert = [UIAlertController alertControllerWithTitle:NSLocalizedString(@"Bad login",nil) message:errorMessage preferredStyle:UIAlertControllerStyleAlert];
			[alert addAction:[UIAlertAction actionWithTitle:NSLocalizedString(@"OK",nil) style:UIAlertActionStyleCancel handler:nil]];
			[self presentViewController:alert animated:YES completion:nil];
		} else {
			// verifying credentials may update the appDelegate values when we subsitute name for correct case:
			_username.text	= username;
			_password.text	= password;
			[_username resignFirstResponder];
			[_password resignFirstResponder];
            
            [self saveVerifiedCredentialsWithUsername:username password:password];

			UIAlertController * alert = [UIAlertController alertControllerWithTitle:NSLocalizedString(@"Login successful",nil) message:nil preferredStyle:UIAlertControllerStyleAlert];
			[alert addAction:[UIAlertAction actionWithTitle:NSLocalizedString(@"OK",nil) style:UIAlertActionStyleCancel handler:^(UIAlertAction * _Nonnull action) {
				[self.navigationController popToRootViewControllerAnimated:YES];
			}]];
			[self presentViewController:alert animated:YES completion:nil];
		}
	}];
}

- (void)viewWillAppear:(BOOL)animated
{
	[super viewWillAppear:animated];

	AppDelegate * appDelegate = (id)[[UIApplication sharedApplication] delegate];
	_username.text	= appDelegate.userName;
	_password.text	= appDelegate.userPassword;

	_saveButton.enabled = _username.text.length && _password.text.length;
}

#pragma mark - Table view delegate

@end

File: src/iOS/LoginViewController.swift
//
//  LoginViewController.swift
//  Go Map!!
//
//  Created by Bryce Cogswell on 12/19/12.
//  Copyright (c) 2012 Bryce Cogswell. All rights reserved.
//

import UIKit

class LoginViewController: UITableViewController {
	@IBOutlet var saveButton: UIBarButtonItem!
	@IBOutlet var activityIndicator: UIActivityIndicatorView!

	override func viewDidLoad() {
		super.viewDidLoad()
	}

	@IBAction func registerAccount(_ sender: Any) {
		if let url = URL(string: "https://www.openstreetmap.org/user/new") {
			UIApplication.shared.open(
				url,
				options: [:],
				completionHandler: nil)
		}
	}

	@IBAction func verifyAccount(_ sender: Any) {
		if activityIndicator.isAnimating {
			return
		}

		activityIndicator.color = UIColor.darkGray
		activityIndicator.startAnimating()

		AppDelegate.shared.oAuth2.getUserDetails(callback: { dict in
			self.activityIndicator.stopAnimating()

			if let dict = dict {
				if let name = dict["display_name"] as? String {
					AppDelegate.shared.userName = name
				}
				let alert = UIAlertController(
					title: NSLocalizedString("Login successful", comment: ""),
					message: nil,
					preferredStyle: .alert)
				alert.addAction(UIAlertAction(title: NSLocalizedString("OK", comment: ""),
				                              style: .cancel,
				                              handler: { _ in
				                              	self.navigationController?.popToRootViewController(animated: true)
				                              }))
				self.present(alert, animated: true)
			} else {
				let alert = UIAlertController(
					title: NSLocalizedString("Bad login", comment: ""),
					message: NSLocalizedString("Not found", comment: "User credentials not found"),
					preferredStyle: .alert)
				alert.addAction(UIAlertAction(title: NSLocalizedString("OK", comment: ""),
				                              style: .cancel,
				                              handler: nil))
				self.present(alert, animated: true)
			}
		})
	}

	override func viewWillAppear(_ animated: Bool) {
		super.viewWillAppear(animated)

		saveButton.isEnabled = AppDelegate.shared.oAuth2.isAuthorized()
	}

	@IBAction func loginWithOAuth(_ sender: Any?) {
		AppDelegate.shared.oAuth2.requestAccessFromUser(withVC: self, onComplete: { result in
			switch result {
			case .success:
				let alert = UIAlertController(
					title: NSLocalizedString("Login successful", comment: ""),
					message: nil,
					preferredStyle: .alert)
				alert.addAction(UIAlertAction(title: NSLocalizedString("OK", comment: ""),
				                              style: .cancel,
				                              handler: { _ in
				                              	self.navigationController?.popToRootViewController(animated: true)
				                              }))
				self.present(alert, animated: true)
			case let .failure(error):
				let alert = UIAlertController(
					title: NSLocalizedString("Bad login", comment: ""),
					message: error.localizedDescription,
					preferredStyle: .alert)
				alert.addAction(UIAlertAction(title: NSLocalizedString("OK", comment: ""),
				                              style: .cancel,
				                              handler: nil))
				self.present(alert, animated: true)
			}
		})
	}
}
